federation/pkg/federation-controller/service/service_helper.go

```diff
@@ -0,0 +1,219 @@
+	cachedStatus, ok := cachedService.serviceStatusMap[clusterName]
+	// cached status found, remove ingress info from federation service
+	if ok {
+		// new status = cached federation status - cached status for the cluster + new status from the cluster
+		cachedFedServiceStatus := cachedService.lastState.Status.LoadBalancer
+		for i, fed := range cachedFedServiceStatus.Ingress {
+			for _, old := range cachedStatus.Ingress{
+				// remove if same ingress record found
+				if (old.IP == fed.IP && old.Hostname == fed.Hostname) {
+					cachedFedServiceStatus.Ingress = append(cachedFedServiceStatus.Ingress[:i], cachedFedServiceStatus.Ingress[i+1:]...)
+					glog.V(4).Infof("remove old ingress info '%s', '%s' for service %s/%s", old.Hostname, old.IP, cachedService.lastState.Namespace, cachedService.lastState.Name)
+				}
+			}
+		}
+		delete(cachedService.serviceStatusMap, clusterName)
+		delete(cachedService.endpointMap, clusterName)
+		return true
+	} else {
+		glog.V(4).Infof("service removal %s/%s from cluster %s observed, but cached endpoint info is not changed.", cachedService.lastState.Namespace, cachedService.lastState.Name, clusterName)
+	}
+	return false
+}
+
+// Update ingress info when service updated, and persist to federation apiserver
+func (cc *clusterClientCache) processServiceUpdate(cachedService *cachedService, service *v1.Service, clusterName string) bool {
+	glog.V(4).Infof("Processing service update for %s/%s, cluster %s", service.Namespace, service.Name, clusterName)
+	cachedService.rwlock.Lock()
+	defer cachedService.rwlock.Unlock()
+	var needUpdate bool
+	cachedStatus, ok := cachedService.serviceStatusMap[clusterName]
+	cachedFedServiceStatus := cachedService.lastState.Status.LoadBalancer
+	if ok {
+		if reflect.DeepEqual(cachedStatus, service.Status.LoadBalancer) {
+			glog.V(4).Infof("same ingress info observed for service %s/%s: %+v ", service.Namespace, service.Name, cachedStatus.Ingress)
+		} else{
+			glog.V(4).Infof("ingress info was changed for service %s/%s: %+v ", service.Namespace, service.Name, cachedStatus.Ingress)
+			needUpdate = true
+		}
+	} else {
+		glog.V(4).Infof("cached service status was not found for %s/%s, cluster %s, building one", service.Namespace, service.Name, clusterName)
+		cachedService.serviceStatusMap[clusterName] = service.Status.LoadBalancer
+		needUpdate = true
+
+		// cache is not always reliable(cache will be cleaned when service controller restart)
+		// two cases will run into this branch:
+		// 1. new service loadbalancer info received -> no info in cache, and no in federation service
+		// 2. service controller being restarted -> no info in cache, but it is in federation service
+
+		// check if the lb info is already in federation service
+		CheckLB:
+			for _, fed := range cachedFedServiceStatus.Ingress {
+				for _, new := range service.Status.LoadBalancer.Ingress{
+					if (new.IP == fed.IP && new.Hostname == fed.Hostname) {
+						needUpdate = false
```

Comment:
I don't think that this logic works correctly.  You need an update if any of the ingresses are different, right?. As the code stands, I think that it will return needUpdate = false if it finds any single ingress that matches between fed and new?   I think that you might need to reverse the logic, starting with the assumption that needUpdate = false, and setting it to true as soon as you find a non-matching ingress (just outside the inner for loop)?